File: ios/Tiny/Sources/AgentMap.swift
/**
 * 🗺️ Agent map bridge — the iOS third of web's __tinyMapBridge and
 * Android's AgentMap.kt (agi-diy index.html:2607 lineage). The chat stream
 * feeds map tool calls here; TinyMapView and the ambient map observe and
 * draw. Pins are keyed by the id the AGENT chose (re-using an id moves that
 * pin), so remove/fly-to/tour need no round-trip — the model references ids
 * it assigned. State outlives any one map view on purpose: pins placed
 * while the map is hidden greet the user when 📍 goes on.
 *
 * Foundation-only math/parsing lives in nonisolated statics so TinyTests
 * can pin the clamps without MapKit or a MainActor hop.
 */
import Foundation
import Combine

@MainActor
final class AgentMap: ObservableObject {
    static let shared = AgentMap()

    struct Pin: Identifiable, Equatable {
        let id: String
        let lat: Double
        let lng: Double
        let label: String?
        /// raw CSS-ish hex from the tool call; color-mapped at render time
        let color: String?
    }

    /// One camera gesture; `seq` makes each command a distinct Equatable
    /// value so .onChange fires even for a repeated destination.
    struct Camera: Equatable {
        let seq: Int
        let lat: Double
        let lng: Double
        let zoom: Double?
    }

    @Published private(set) var pins: [String: Pin] = [:]
    @Published private(set) var camera: Camera?

    private var autoId = 0
    private var seq = 0
    private var tourTask: Task<Void, Never>?

    // Mounted map surfaces (ambient/full-screen) check in so the chat can
    // tell when a map gesture would be INVISIBLE and hint instead (web
    // parity: the "tap 📍 to see it" toast).
    @Published private(set) var visibleMaps = 0
    var mapVisible: Bool { visibleMaps > 0 }
    func mapDidAppear() { visibleMaps += 1 }
    func mapDidDisappear() { visibleMaps = max(0, visibleMaps - 1) }

    /// Returns true when the tool name was a map tool (handled here).
    @discardableResult
    func handle(name: String, argsJson: String) -> Bool {
        let input = (try? JSONSerialization.jsonObject(with: Data(argsJson.utf8))) as? [String: Any] ?? [:]
        switch name {
        case "add_map_marker":
            guard let (lat, lng) = Self.coords(input) else { return true }
            var id = (input["id"] as? String)?.trimmingCharacters(in: .whitespaces) ?? ""
            if id.isEmpty {
                autoId += 1
                id = "pin-\(autoId)"
            }
            id = String(id.prefix(32))
            let label = (input["label"] as? String).flatMap { $0.isEmpty ? nil : String($0.prefix(40)) }
            let color = (input["color"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            pins[id] = Pin(id: id, lat: lat, lng: lng, label: label, color: color)
        case "remove_map_marker":
            if let p = Self.resolvePin(input["id"] as? String ?? "", in: pins) {
                pins.removeValue(forKey: p.id)
            }
        case "clear_map_markers":
            clearPins()
        case "fly_to_location":
            if let (lat, lng) = Self.coords(input) { fly(lat: lat, lng: lng, zoom: Self.zoom(input)) }
        case "fly_to_marker":
            if let p = Self.resolvePin(input["id"] as? String ?? "", in: pins) {
                fly(lat: p.lat, lng: p.lng, zoom: Self.zoom(input))
            }
        case "tour_markers":
            let stops = Self.tourStops(input, pins: pins)
            guard !stops.isEmpty else { return true }
            let pause = Self.tourPauseMs(input)
            tourTask?.cancel()
            tourTask = Task { [weak self] in
                for p in stops {
                    guard let self, !Task.isCancelled else { return }
                    self.fly(lat: p.lat, lng: p.lng, zoom: nil)
                    try? await Task.sleep(nanoseconds: UInt64(pause) * 1_000_000)
                }
            }
        default:
            return false
        }
        return true
    }

    /// When the agent last steered the camera — the ambient follow loop
    /// yields for 30s after a gesture instead of stomping it on its next
    /// 15s tick (fly_to_marker looked broken: the map snapped home).
    private(set) var lastGestureAt: Date?
    var followSuspended: Bool {
        guard let t = lastGestureAt else { return false }
        return Date().timeIntervalSince(t) < 30
    }

    /// Spotlight: while the agent is presenting (fly/tour), the ambient
    /// wash thins so the map comes FORWARD, then fades back. Duration is
    /// settable for tests.
    @Published private(set) var spotlight = false
    var spotlightSeconds: Double = 8
    private var spotlightTask: Task<Void, Never>?

    private func fly(lat: Double, lng: Double, zoom: Double?) {
        seq += 1
        lastGestureAt = Date()
        camera = Camera(seq: seq, lat: lat, lng: lng, zoom: zoom)
        spotlight = true
        spotlightTask?.cancel()
        spotlightTask = Task { [weak self] in
            guard let self else { return }
            try? await Task.sleep(nanoseconds: UInt64(self.spotlightSeconds * 1_000_000_000))
            if !Task.isCancelled { self.spotlight = false }
        }
    }

    /// Drop every agent pin and stop a running tour — the agent's
    /// clear_map_markers and the map screen's "clear pins" button.
    func clearPins() {
        tourTask?.cancel()
        tourTask = nil
        pins = [:]
    }

    /// Tests only — the singleton outlives XCTest cases.
    func resetForTest() {
        tourTask?.cancel()
        tourTask = nil
        spotlightTask?.cancel()
        spotlightTask = nil
        spotlight = false
        spotlightSeconds = 8
        pins = [:]
        camera = nil
        autoId = 0
        seq = 0
        visibleMaps = 0
        lastGestureAt = nil
    }

    // MARK: - pure parsing/math (web schema clamp parity)

    nonisolated static func coords(_ input: [String: Any]) -> (Double, Double)? {
        guard let lat = (input["lat"] as? NSNumber)?.doubleValue,
              let lng = (input["lng"] as? NSNumber)?.doubleValue,
              lat.isFinite, lng.isFinite,
              (-90.0...90.0).contains(lat), (-180.0...180.0).contains(lng) else { return nil }
        return (lat, lng)
    }

    nonisolated static func zoom(_ input: [String: Any]) -> Double? {
        guard let z = (input["zoom"] as? NSNumber)?.doubleValue, z.isFinite else { return nil }
        return min(20, max(1, z))
    }

    nonisolated static func tourPauseMs(_ input: [String: Any]) -> Int {
        let ms = (input["pause_ms"] as? NSNumber)?.intValue ?? 2000
        return min(10_000, max(500, ms))
    }

    /// id → pin, with a LABEL fallback: the model often skips the optional
    /// id on add_map_marker (the auto "pin-N" is never echoed back — these
    /// tools are fire-and-forget), then references the pin by its label.
    /// "fly to the coffee pin" must fly. Exact id wins; else the first
    /// case-insensitive label match.
    nonisolated static func resolvePin(_ ref: String, in pins: [String: Pin]) -> Pin? {
        if let p = pins[ref] { return p }
        let needle = ref.trimmingCharacters(in: .whitespaces).lowercased()
        guard !needle.isEmpty else { return nil }
        return pins.values.first { ($0.label ?? "").lowercased() == needle }
    }

    /// Known pins in tour order — unknown refs skipped, capped at 12.
    nonisolated static func tourStops(_ input: [String: Any], pins: [String: Pin]) -> [Pin] {
        let ids = (input["ids"] as? [Any])?.compactMap { $0 as? String } ?? []
        return Array(ids.compactMap { resolvePin($0, in: pins) }.prefix(12))
    }

    /// Web zoom level → MapKit span degrees (360° at zoom 0, halved per
    /// level — the Mercator convention agi-diy zooms mean). nil zoom keeps
    /// the streets-ish default the follow loops use.
    nonisolated static func spanDegrees(forZoom zoom: Double?) -> Double {
        guard let zoom else { return 0.02 }
        return 360.0 / pow(2.0, min(20, max(1, zoom)))
    }

    /// #rgb/#rrggbb → 0…1 components; nil for junk (render falls back to
    /// the tiny's accent). Pure math — usable off MainActor and in tests.
    nonisolated static func rgb(fromHex color: String?) -> (r: Double, g: Double, b: Double)? {
        guard var hex = color?.trimmingCharacters(in: .whitespaces), hex.hasPrefix("#") else { return nil }
        hex.removeFirst()
        let parts: [String]
        switch hex.count {
        case 3: parts = hex.map { "\($0)\($0)" }
        case 6: parts = stride(from: 0, to: 6, by: 2).map {
            let s = hex.index(hex.startIndex, offsetBy: $0)
            return String(hex[s..<hex.index(s, offsetBy: 2)])
        }
        default: return nil
        }
        let vals = parts.compactMap { Int($0, radix: 16) }
        guard vals.count == 3 else { return nil }
        return (Double(vals[0]) / 255, Double(vals[1]) / 255, Double(vals[2]) / 255)
    }
}

File: android/app/src/main/java/technology/tiny/app/tools/AgentMap.kt
package technology.tiny.app.tools

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject

/**
 * 🗺️ Agent map bridge — the Android half of the web `__tinyMapBridge`
 * (agi-diy index.html:2607 port). ChatViewModel feeds beforeToolCallEvent
 * map tools here; MapBackdrop/MapSheet observe the flows and draw. Pins are
 * keyed by the id the AGENT chose (re-using an id moves that pin), so
 * remove/fly-to/tour need no round-trip: the model references ids it
 * assigned. State outlives any one map composable on purpose — pins placed
 * while the map is off greet the user when they enable 📍.
 */
object AgentMap {
    data class Pin(
        val id: String,
        val lat: Double,
        val lng: Double,
        val label: String? = null,
        /** raw CSS-ish hex from the tool call; hue-mapped at render time */
        val color: String? = null,
    )

    /** One camera gesture; `seq` makes each command distinct so collectors
     *  can skip the stale latest value on (re)mount. */
    data class Camera(val seq: Long, val lat: Double, val lng: Double, val zoom: Float? = null)

    private val _pins = MutableStateFlow<Map<String, Pin>>(emptyMap())
    val pins: StateFlow<Map<String, Pin>> get() = _pins

    private val _camera = MutableStateFlow<Camera?>(null)
    val camera: StateFlow<Camera?> get() = _camera

    private var autoId = 0L
    private var seq = 0L
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var tourJob: Job? = null

    // Composed map surfaces (backdrop/sheet) check in so the chat can tell
    // when a map gesture would be INVISIBLE and hint instead (web parity:
    // the "tap 📍 to see it" toast).
    private val _visibleMaps = MutableStateFlow(0)
    val mapVisible: Boolean get() = _visibleMaps.value > 0
    fun mapShown() { _visibleMaps.value += 1 }
    fun mapHidden() { _visibleMaps.value = (_visibleMaps.value - 1).coerceAtLeast(0) }

    /** Returns true when the tool name was a map tool (handled here). */
    fun handle(name: String, input: JSONObject): Boolean {
        when (name) {
            "add_map_marker" -> {
                val lat = input.optDouble("lat", Double.NaN)
                val lng = input.optDouble("lng", Double.NaN)
                if (!validCoords(lat, lng)) return true
                val id = input.optString("id").ifBlank { "pin-${++autoId}" }.take(32)
                _pins.value = _pins.value + (id to Pin(
                    id = id,
                    lat = lat,
                    lng = lng,
                    label = input.optString("label").ifBlank { null }?.take(40),
                    color = input.optString("color").ifBlank { null },
                ))
            }
            "remove_map_marker" -> {
                resolvePin(input.optString("id"))?.let { _pins.value = _pins.value - it.id }
            }
            "clear_map_markers" -> clearPins()
            "fly_to_location" -> {
                val lat = input.optDouble("lat", Double.NaN)
                val lng = input.optDouble("lng", Double.NaN)
                if (validCoords(lat, lng)) fly(lat, lng, zoomOrNull(input))
            }
            "fly_to_marker" -> {
                resolvePin(input.optString("id"))?.let { fly(it.lat, it.lng, zoomOrNull(input)) }
            }
            "tour_markers" -> {
                val stops = tourStops(input)
                if (stops.isEmpty()) return true
                val pause = tourPauseMs(input)
                tourJob?.cancel()
                tourJob = scope.launch {
                    for (p in stops) {
                        fly(p.lat, p.lng, null)
                        delay(pause)
                    }
                }
            }
            else -> return false
        }
        return true
    }

    /** Drop every agent pin and stop a running tour — the agent's
     *  clear_map_markers and the map screens' "clear pins" button. */
    fun clearPins() {
        tourJob?.cancel()
        tourJob = null
        _pins.value = emptyMap()
    }

    /** id → pin, with a LABEL fallback (iOS resolvePin parity): the model
     *  often skips the optional id on add (the auto "pin-N" is never echoed
     *  back), then references the pin by its label — "fly to the coffee
     *  pin" must fly. Exact id wins; else first case-insensitive label. */
    internal fun resolvePin(ref: String): Pin? {
        _pins.value[ref]?.let { return it }
        val needle = ref.trim().lowercase()
        if (needle.isEmpty()) return null
        return _pins.value.values.firstOrNull { (it.label ?: "").lowercase() == needle }
    }

    /** Known pins in tour order — unknown refs skipped, capped at 12 (web bridge parity). */
    internal fun tourStops(input: JSONObject): List<Pin> {
        val ids = input.optJSONArray("ids") ?: return emptyList()
        return (0 until ids.length()).mapNotNull { resolvePin(ids.optString(it)) }.take(12)
    }

    internal fun tourPauseMs(input: JSONObject): Long =
        input.optLong("pause_ms", 2000L).coerceIn(500L, 10_000L)

    private fun validCoords(lat: Double, lng: Double): Boolean =
        lat.isFinite() && lng.isFinite() && lat in -90.0..90.0 && lng in -180.0..180.0

    private fun zoomOrNull(input: JSONObject): Float? {
        val z = input.optDouble("zoom", Double.NaN)
        return if (z.isFinite()) z.toFloat().coerceIn(1f, 20f) else null
    }

    /** When the agent last steered the camera (epoch ms — System.currentTimeMillis
     *  is JVM-test-safe, SystemClock isn't). The ambient follow loop yields for
     *  30s after a gesture instead of stomping it on its next 15s tick
     *  (fly_to_marker looked broken: the map snapped home). */
    @Volatile
    var lastGestureAtMs = 0L
        private set
    val followSuspended: Boolean
        get() = System.currentTimeMillis() - lastGestureAtMs < 30_000

    /** Spotlight: while the agent is presenting (fly/tour), the chat wash
     *  thins so the map comes FORWARD, then fades back. Duration settable
     *  for tests. */
    private val _spotlight = MutableStateFlow(false)
    val spotlight: StateFlow<Boolean> get() = _spotlight
    internal var spotlightMs = 8_000L
    private var spotlightJob: Job? = null

    @Synchronized
    private fun fly(lat: Double, lng: Double, zoom: Float?) {
        lastGestureAtMs = System.currentTimeMillis()
        _camera.value = Camera(++seq, lat, lng, zoom)
        _spotlight.value = true
        spotlightJob?.cancel()
        spotlightJob = scope.launch {
            delay(spotlightMs)
            _spotlight.value = false
        }
    }

    /**
     * Marker hue (0..360) for a #rgb/#rrggbb color, null → default red pin.
     * Pure rgb→hue math — android.graphics.Color is unavailable on the JVM.
     */
    fun markerHue(color: String?): Float? {
        val hex = color?.trim()?.removePrefix("#") ?: return null
        val rgb = when (hex.length) {
            3 -> hex.map { "$it$it".toIntOrNull(16) ?: return null }
            6 -> hex.chunked(2).map { it.toIntOrNull(16) ?: return null }
            else -> return null
        }
        val (r, g, b) = rgb.map { it / 255f }
        val max = maxOf(r, g, b)
        val min = minOf(r, g, b)
        val d = max - min
        if (d == 0f) return null // greys have no hue — keep the default pin
        val h = when (max) {
            r -> ((g - b) / d) % 6f
            g -> (b - r) / d + 2f
            else -> (r - g) / d + 4f
        } * 60f
        return if (h < 0f) h + 360f else h
    }

    /** Tests only — the singleton outlives JUnit classes. */
    internal fun resetForTest() {
        tourJob?.cancel()
        tourJob = null
        spotlightJob?.cancel()
        spotlightJob = null
        _spotlight.value = false
        spotlightMs = 8_000L
        _pins.value = emptyMap()
        _camera.value = null
        _visibleMaps.value = 0
        autoId = 0L
        seq = 0L
        lastGestureAtMs = 0L
    }
}
